src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,594 @@
+ * <p>
+ * API Usage Example:
+ * {@snippet lang = java:
+ *    KDF kdfHkdf = KDF.getInstance("HKDFWithHmacSHA256");
+ *
+ *    AlgorithmParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 42);
+ *
+ *    kdfHkdf.deriveKey("AES", kdfParameterSpec);
+ *}
+ *
+ * @see SecretKey
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KDF");
+
+    private static final Debug pdebug = Debug.getInstance("provider",
+                                                          "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=")
+                                             && !Debug.isOn("kdf");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the KDF algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final KDFParameters kdfParameters;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KDF object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param kdfParameters
+     *     the algorithm parameters
+     */
+    private KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                KDFParameters kdfParameters) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        lock = new Object();
+    }
+
+    private KDF(Service s, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        firstService = s;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        chooseFirstProvider();
+        return provider.getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used to initialize the object.
+     *
+     * @return the parameters used to initialize the object
```

Comment:
Can this be null if not provided or provided as null? Is that possible that even if there is no parameters provided but the impl might choose default parameters and return them here?